x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Bucket.java

```diff
@@ -336,9 +333,18 @@ protected TypeResolution resolveType() {
                 : resolution.and(checkArgsCount(2)); // temporal amount
         }
         if (fieldType.isNumeric()) {
-            return bucketsType.isWholeNumber()
-                ? checkArgsCount(4).and(() -> isNumeric(from, sourceText(), THIRD)).and(() -> isNumeric(to, sourceText(), FOURTH))
-                : isNumeric(buckets, sourceText(), SECOND).and(checkArgsCount(2));
+            return isNumeric(buckets, sourceText(), SECOND).and(() -> {
+                if (bucketsType.isRationalNumber()) {
+                    return checkArgsCount(2);
+                } else { // second arg is a whole number: either a span, but as a whole, or count, and we must expect a range
+                    var resolution = checkArgsCount(2);
+                    if (resolution.resolved() == false) {
+                        resolution = checkArgsCount(4).and(() -> isNumeric(from, sourceText(), THIRD))
+                            .and(() -> isNumeric(to, sourceText(), FOURTH));
```

Comment:
Do we have (verifier) test cases for when the args count is higher than 4, or the 3rd and 4th are not numeric?

Reply:
Added some.
Not for a higher than 4 count, though, since this would be a general parser exception, raised when resolving the function.